### PR TITLE
Exception raised for hog surprising result

### PR DESCRIPTION
**Pull request by farrjere**
…ameters given with hog

## Description
Addresses #7064 by throwing an exception when an image is too small for the parameters given, instead of returning an empty array. 

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
...
Added in exception to avoid surprising result when image is too small for the given parameters of hog
```

### CODE REVIEW

File: skimage/feature/_hog.py [hunk 1]
```diff
@@ -276,6 +282,9 @@ def hog(image, orientations=9, pixels_per_cell=(8, 8), cells_per_block=(3, 3),
 
     n_blocks_row = (n_cells_row - b_row) + 1
     n_blocks_col = (n_cells_col - b_col) + 1
+    if n_blocks_col <= 0 or n_blocks_row <= 0:
+        raise ValueError('The image given is not large enough for the cells_per_block '
+                         'and pixels_per_cell given')
```
Comment (soupault) [inline skimage/feature/_hog.py:287]: What do you think about telling to the user in the same error message what should be the minimal size of the image given the provided `cells_per_block` and `pixels_per_cell`?

File: skimage/feature/tests/test_hog.py [hunk 2]
```diff
@@ -287,3 +287,16 @@ def test_hog_output_equivariance_channel_axis(channel_axis):
...
+
+def test_hog_small_image():
+    """
+    tests that an exception is thrown instead of an empty array returned when
+    the image is smaller than it should be for the given parameters
+    """
+    img = np.zeros((24, 193))
+    feature.hog(img)
+
+    img = np.zeros((15, 193))
```
Comment (soupault) [inline skimage/feature/tests/test_hog.py:300]: I think it would be more future-compatible to explicitly specify the values of `pixels_per_cell` and `cells_per_block` parameters in this test.

Comment (farrjere) [reply skimage/feature/tests/test_hog.py:300]: That makes sense I have updated this accordingly.

File: skimage/feature/_hog.py (hunk 1 above)
Comment (farrjere) [reply skimage/feature/_hog.py]: I agree that would be more clear, I have updated this accordingly.

File: skimage/feature/_hog.py [hunk 3]
```diff
@@ -276,6 +282,12 @@ def hog(image, orientations=9, pixels_per_cell=(8, 8), cells_per_block=(3, 3),
 
     n_blocks_row = (n_cells_row - b_row) + 1
     n_blocks_col = (n_cells_col - b_col) + 1
+    if n_blocks_col <= 0 or n_blocks_row <= 0:
+        min_row = (-1 + b_row) * c_row + 1
+        min_col = (-1 + b_col) * c_col + 1
+        raise ValueError('The image given is not large enough for the cells_per_block '
+                         'and pixels_per_cell given, should be at least:'
+                         f' {min_row} rows x {min_col} cols')
```
Comment (lagru) [inline skimage/feature/_hog.py:290]: I think this can't be right. For 
```python
import numpy as np
import skimage
img = np.zeros((17, 17))
skimage.feature.hog(img)
```
it still get the error telling me `should be at least: 17 rows x 17 cols` which `img` satisfies...

File: skimage/feature/tests/test_hog.py [hunk 4]
```diff
@@ -287,3 +287,16 @@ def test_hog_output_equivariance_channel_axis(channel_axis):
...
         assert_almost_equal(hog_ref, hog_fact)
+
+
+def test_hog_small_image():
+    """
+    tests that an exception is thrown instead of an empty array returned when
+    the image is smaller than it should be for the given parameters
+    """
+    img = np.zeros((24, 193))
+    feature.hog(img, pixels_per_cell=(8, 8), cells_per_block=(3, 3))
```
Comment (lagru) [inline skimage/feature/tests/test_hog.py:298]: Could you use the minimal required image size here? That way we can be sure that future changes won't affect the edge case. If you want you could even pass in different `pixels_per_cell` and `cells_per_block` with `@pytest.mark.parametrize` and calculate the passing and failing image size on the fly...

File: skimage/feature/tests/test_hog.py [hunk 5]
```diff
@@ -287,3 +287,16 @@ def test_hog_output_equivariance_channel_axis(channel_axis):
...
+
+    img = np.zeros((15, 193))
+    with pytest.raises(ValueError):
```
Comment (lagru) [inline skimage/feature/tests/test_hog.py:301]: Could you use the `match` keyword here to ensure that the correct value error is raised? You don't need to match the entire message, e.g. 
```suggestion
    with pytest.raises(ValueError, match=".*image given is not large enough"):
```

**Comment by lagru** [comment]
Thanks for tackling this @farrjere! :D

File: skimage/feature/tests/test_hog.py [hunk 6]
```diff
@@ -287,3 +287,16 @@ def test_hog_output_equivariance_channel_axis(channel_axis):
...
+def test_hog_small_image():
+    """
+    tests that an exception is thrown instead of an empty array returned when
+    the image is smaller than it should be for the given parameters
+    """
+    img = np.zeros((24, 24))
+    feature.hog(img, pixels_per_cell=(8, 8), cells_per_block=(3, 3))
+
+    img = np.zeros((23, 23))
+    with pytest.raises(ValueError, match=".*image given is not large enough"):
```
Comment (mkcor) [inline skimage/feature/tests/test_hog.py:301]:
```suggestion
    with pytest.raises(ValueError, match=".*image is too small given"):
```

**Review by mkcor — COMMENTED** [review]
@farrjere thank you so much for this first contribution of yours!

**Review by lagru — APPROVED** [review]
LGTM, except for a few open minor suggestions. Thanks.

**Comment by mkcor** [comment]
@farrjere seeing bfcacc6fcf84cf4c9c852d18a289dfb222b40cf0, note that you may commit suggestions directly through the web interface (then sync up your branch locally if needed). When there are several suggestions, you may also commit multiple at once: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

**Comment by mkcor** [comment]
@farrjere fyi I've applied the remaining suggestions, which I believe shouldn't be controversial. I'll wait for CI to complete and pass. Thanks again for this contribution!

**Comment by farrjere** [comment]
Thank you I wouldn't have been able to get to it until tonight, so I appreciate it.

**Comment by lagru** [comment]
@farrjere thanks for your first contribution! :D
